lib/ansible/vars/manager.py

```diff
@@ -388,18 +383,15 @@ def plugins_by_groups():
                 raise AnsibleParserError("Error while reading vars files - please supply a list of file names. "
                                          "Got '%s' of type %s" % (vars_files, type(vars_files)))
 
-            # We now merge in all exported vars from all roles in the play,
-            # unless the user has disabled this
-            # role is public and
-            #    either static or dynamic and completed
-            # role is not set
-            #    use config option as default
-            for role in play.get_roles():
-                role_is_static_or_completed = role.static or role._completed.get(host.name, False)
-                if role.public and role_is_static_or_completed or \
-                   role.public is None and not C.DEFAULT_PRIVATE_ROLE_VARS and role_is_static_or_completed:
-
-                    all_vars = _combine_and_track(all_vars, role.get_vars(include_params=False, only_exports=True), "role '%s' exported vars" % role.name)
+            # We now merge in all exported vars from public roles in the play
+            if host:
+                for role in play.get_roles():
+                    if role.static or role._completed.get(host.name, False):
```

Comment:
@s-hertel @bcoca I am also confused about the requirement for dynamic roles to be completed before their variables are exposed to the play. If I am reading this correctly it means that the dynamic role variables are exposed to the play only for the hosts where the dynamic role was completed.

So there is a difference in behavior for static and dynamic roles in that dynamic role variables are exposed to the play only when the role completed and only on the hosts on which it completed. As opposed to static roles where role variables are exposed to the play immediately and to all hosts. Is that correct?

But we do not seem to document it.

Reply:
I don't see an obvious downside to removing the `role._completed` check. Just removing that bit was my preferred fix for https://github.com/ansible/ansible/issues/80944. I'm not totally sure what the original intention of the check was since it applied to import_role initially.

Reply:
Since `include` has never actually relied on `role._completed` because of bugs (first in 2.15 only applied to `import_role`, now on devel doesn't matter because `static` is always True), wouldn't this be the backwards incompatible change?

It seems like a complicated use-case, but if it was toggle-able (and supported by both include_role and import_role), that would seem better to me. Wasn't there a conversation about making `public` accept a 3rd option? `public: true|false|on_host_complete`.

Reply:
that was for imports, which i had 'broken' to be on completion in previous PR and had to fix, which is how we got to current code, so imports are always updated while includes were supposed to only be updated after they run. Both obeying public/private now.

Reply:
Yes. I'm saying it's a change in behavior for include_role too, and why not make it optional.

Reply:
because then `true` would have to update the vars before we know if the include itself happens.

Reply:
> Both obeying public/private now.

Just to clarify, I wasn't talking about obeying private at all. `include_role` public vars haven't been restricted to hosts that finished the role before. My thought was since it never worked, we could make the UI more flexible or remove without breaking playbooks.

Reply:
That was the 'bug' i fixed for imports, vars were not appearing until role completion, but if that fix also added includes .. that was a bug I introduced and not how it was meant to work.

Reply:
Makes sense. Hopefully the fix will allow a way to restore the long-term behavior, as I can see people wanting include_role vars available on all hosts or the individual host that finished the role.

Reply:
also ..  current code is doing 'per host per role' if im reading it correctly so not sure if we already break the use case I gave above. Unless the way we mark completed per role/host can affect hosts that didnt run that specific role (skipping counts?).

Reply:
i thought we fixed it so all roles do get a completed now, no matter the task content, need to x2 check that. 

Yes, the old behavior should be the `run_once: true` equivalent, my branch plays with keeping it per host, but I'm still unsure if we should keep the previous handling and how/when/if we deal with conflicts (which we currently resolve w/o specifying):  

 - hosta uses roleb, hostb uses rolea for same include and each role has diff set of vars, but some overlap
 - hosta and hostb use rolea but both update the vars in diff ways
 - hosta uses rolea but hostb does not

And those are the simple ones .. we can add more hosts and roles.

Reply:
Instead of adding the host to play.roles() on completed, putting it back in when loading the block list will restore 'run_once: true'

Reply:
the eor task is now unconditionally added to the task list, even if empty:
https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role/__init__.py#L609

and we both compile roles at play start and on include/import

Reply:
> I'm still unsure if we should keep the previous handling and how/when/if we deal with conflicts (which we currently resolve w/o specifying):

I'm not sure we need to do anything special with conflicts :thinking: Maybe some playbooks are hash merging the exported vars... But generally last added to play roles wins. Tasks in roles prefer their own inheritance/dep chain after the play roles.

We do always inject a meta role_complete task but only mark the role as complete if it had a task. This behavior was differently briefly, and broke conditionally including role dependencies https://github.com/ansible/ansible/issues/81486.

Reply:
ah, we did that on purpose, didn't realize it created issues and was reverted